File: Examples/Advanced/Export-Device-Inventory/Python/export_device_inventory.py
#!/usr/bin/env python3
"""Export the complete netLD/ThirdEye device inventory to CSV or JSON."""

from __future__ import annotations

import argparse
import csv
import json
import os
import sys
import tempfile
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Iterator

import requests
from dotenv import load_dotenv


CSV_FIELDS = (
    "network", "ipAddress", "hostname", "adapterId", "deviceType",
    "hardwareVendor", "model", "serialNumber", "softwareVendor", "osVersion",
    "backupStatus", "complianceState", "lastBackup", "lastTelemetry", "memoSummary",
    "custom1", "custom2", "custom3", "custom4", "custom5",
)


class ExampleError(RuntimeError):
    pass


def parse_networks(value: str) -> list[str]:
    networks = [item.strip() for item in value.split(",") if item.strip()]
    if not networks:
        raise ExampleError("NETLD_NETWORKS must contain at least one managed network.")
    return networks


@dataclass(frozen=True)
class Config:
    base_url: str
    api_key: str
    networks: list[str]
    output_path: Path
    output_format: str
    page_size: int
    scheme: str
    query: str

    @classmethod
    def from_env(
        cls, env_path: Path, output_base: Path | None = None,
        format_override: str | None = None,
    ) -> "Config":
        load_dotenv(env_path, override=True)
        base_url = os.getenv("NETLD_BASE_URL", "").strip()
        api_key = os.getenv("NETLD_API_KEY", "").strip()
        if not base_url:
            raise ExampleError("Set NETLD_BASE_URL in .env before running this example.")
        if not api_key:
            raise ExampleError("Set NETLD_API_KEY in .env before running this example.")
        try:
            page_size = int(os.getenv("NETLD_PAGE_SIZE", "500"))
        except ValueError as exc:
            raise ExampleError("NETLD_PAGE_SIZE must be a positive integer.") from exc
        if page_size <= 0:
            raise ExampleError("NETLD_PAGE_SIZE must be a positive integer.")
        output_format = (format_override or os.getenv("NETLD_OUTPUT_FORMAT", "csv")).strip().lower()
        if output_format not in {"csv", "json"}:
            raise ExampleError("NETLD_OUTPUT_FORMAT must be either csv or json.")
        output_name = os.getenv("NETLD_OUTPUT_FILE", "").strip() or f"inventory.{output_format}"
        output_path = Path(output_name)
        if not output_path.is_absolute():
            output_path = (output_base or env_path.parent) / output_path
        return cls(
            base_url=base_url.rstrip("/"), api_key=api_key,
            networks=parse_networks(os.getenv("NETLD_NETWORKS", "Default")),
            output_path=output_path, output_format=output_format, page_size=page_size,
            scheme=os.getenv("NETLD_SEARCH_SCHEME", "ipAddress").strip() or "ipAddress",
            query=os.getenv("NETLD_SEARCH_QUERY", ""),
        )


class NetLDClient:
    def __init__(self, base_url: str, api_key: str, timeout: float = 30):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Authorization": f"Bearer {api_key}"})

    def _request(self, method: str, **kwargs: Any):
        try:
            response = self.session.request(
                method, f"{self.base_url}/rest", timeout=self.timeout,
                allow_redirects=False, **kwargs,
            )
        except requests.RequestException as exc:
            raise ExampleError(f"Could not reach {self.base_url}.") from exc
        if 300 <= response.status_code < 400:
            raise ExampleError(f"Request redirected to {response.headers.get('Location', '')}.")
        try:
            response.raise_for_status()
        except requests.RequestException as exc:
            raise ExampleError(f"Request failed with HTTP {response.status_code}.") from exc
        return response

    def login(self) -> None:
        self._request("GET")

    def call(self, method: str, **parameters: Any) -> Any:
        response = self._request(
            "POST", headers={"Content-Type": "application/json"},
            json={"jsonrpc": "2.0", "method": method, "params": parameters, "id": str(uuid.uuid4())},
        )
        try:
            data = response.json()
        except ValueError as exc:
            raise ExampleError(f"{method} returned invalid JSON.") from exc
        if data.get("error"):
            raise ExampleError(f"{method} failed: {json.dumps(data['error'])}")
        return data.get("result")

    def search_inventory(
        self, networks: list[str], scheme: str, query: str, offset: int, page_size: int
    ) -> dict[str, Any]:
        if not query.endswith("\n"):
            query = f"{query}\n"
        result = self.call(
            "Inventory.search", network=networks, scheme=scheme, query=query,
            pageData={"offset": offset, "pageSize": page_size},
            sortColumn="ipAddress", descending=False,
        )
        if not isinstance(result, dict):
            raise ExampleError("Inventory.search returned no page data.")
        return result


def inventory_pages(
    client: Any, networks: list[str], scheme: str, query: str, page_size: int
) -> Iterator[list[dict[str, Any]]]:
    offset = 0
    total: int | None = None
    while True:
        page = client.search_inventory(networks, scheme, query, offset, page_size)
        devices = page.get("devices") or []
        if not isinstance(devices, list):
            raise ExampleError("Inventory.search returned an invalid devices collection.")
        yield devices
        returned_page_size = int(page.get("pageSize") or page_size)
        if returned_page_size <= 0:
            raise ExampleError("Inventory.search returned an invalid page size.")
        if total is None and page.get("total") is not None:
            total = int(page["total"])
        if total is not None and offset + len(devices) >= total:
            return
        if total is None and len(devices) < returned_page_size:
            return
        if not devices:
            raise ExampleError("Inventory.search returned an empty page before the reported total.")
        offset += returned_page_size


def export_inventory(
    client: Any, networks: list[str], scheme: str, query: str,
    page_size: int, output_path: Path, output_format: str = "csv",
) -> int:
    output_format = output_format.lower()
    if output_format not in {"csv", "json"}:
        raise ExampleError("Output format must be either csv or json.")
    output_path.parent.mkdir(parents=True, exist_ok=True)
    temporary_path: Path | None = None
    count = 0
    try:
        with tempfile.NamedTemporaryFile(
            mode="w", encoding="utf-8", newline="", prefix=f".{output_path.name}.",
            suffix=".tmp", dir=output_path.parent, delete=False,
        ) as output:
            temporary_path = Path(output.name)
            writer = None
            if output_format == "csv":
                writer = csv.DictWriter(output, fieldnames=CSV_FIELDS, extrasaction="ignore")
                writer.writeheader()
            else:
                output.write("[\n")
            first_json_record = True
            for devices in inventory_pages(client, networks, scheme, query, page_size):
                for device in devices:
                    record = {field: device.get(field) for field in CSV_FIELDS}
                    if writer is not None:
                        writer.writerow(record)
                    else:
                        if not first_json_record:
                            output.write(",\n")
                        rendered = json.dumps(record, indent=2, ensure_ascii=False)
                        output.write("\n".join(f"  {line}" for line in rendered.splitlines()))
                        first_json_record = False
                    count += 1
            if output_format == "json":
                output.write("\n]\n")
        os.replace(temporary_path, output_path)
        return count
    except Exception:
        if temporary_path is not None:
            temporary_path.unlink(missing_ok=True)
        raise


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--env", type=Path, default=Path(__file__).with_name(".env"))
    parser.add_argument("--format", choices=("csv", "json"))
    args = parser.parse_args()
    try:
        config = Config.from_env(args.env, Path(__file__).resolve().parent, args.format)
        client = NetLDClient(config.base_url, config.api_key)
        client.login()
        count = export_inventory(
            client, config.networks, config.scheme, config.query,
            config.page_size, config.output_path, config.output_format,
        )
        print(f"Wrote {count} devices to {config.output_path}")
        return 0
    except (ExampleError, OSError, ValueError) as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: Examples/Advanced/Export-Device-Inventory/nodeJS/export-device-inventory.mjs
#!/usr/bin/env node

import { randomUUID } from "node:crypto";
import { mkdir, open, readFile, rename, rm } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath, pathToFileURL } from "node:url";

export const CSV_FIELDS = [
  "network", "ipAddress", "hostname", "adapterId", "deviceType",
  "hardwareVendor", "model", "serialNumber", "softwareVendor", "osVersion",
  "backupStatus", "complianceState", "lastBackup", "lastTelemetry", "memoSummary",
  "custom1", "custom2", "custom3", "custom4", "custom5",
];

export class ExampleError extends Error {}

export function parseNetworks(value) {
  const networks = value.split(",").map((item) => item.trim()).filter(Boolean);
  if (!networks.length) throw new ExampleError("NETLD_NETWORKS must contain at least one managed network.");
  return networks;
}

export function csvCell(value) {
  if (value == null) return "";
  const text = String(value);
  return /[",\r\n]/.test(text) ? `"${text.replaceAll('"', '""')}"` : text;
}

function csvRow(values) {
  return `${values.map(csvCell).join(",")}\n`;
}

async function loadDotEnv(file) {
  let text;
  try {
    text = await readFile(file, "utf8");
  } catch (error) {
    if (error.code === "ENOENT") return;
    throw error;
  }
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith("#") || !line.includes("=")) continue;
    const offset = line.indexOf("=");
    process.env[line.slice(0, offset).trim()] = line.slice(offset + 1).trim().replace(/^(['"])(.*)\1$/, "$2");
  }
}

async function loadConfig(directory, envFile, formatOverride) {
  await loadDotEnv(envFile);
  for (const name of ["NETLD_BASE_URL", "NETLD_API_KEY"]) {
    if (!process.env[name]?.trim()) throw new ExampleError(`Set ${name} in .env before running this example.`);
  }
  const pageSize = Number.parseInt(process.env.NETLD_PAGE_SIZE || "500", 10);
  if (!Number.isInteger(pageSize) || pageSize <= 0) throw new ExampleError("NETLD_PAGE_SIZE must be a positive integer.");
  const format = (formatOverride || process.env.NETLD_OUTPUT_FORMAT || "csv").trim().toLowerCase();
  if (!["csv", "json"].includes(format)) throw new ExampleError("NETLD_OUTPUT_FORMAT must be either csv or json.");
  return {
    baseUrl: process.env.NETLD_BASE_URL.replace(/\/$/, ""),
    apiKey: process.env.NETLD_API_KEY,
    networks: parseNetworks(process.env.NETLD_NETWORKS || "Default"),
    outputFile: path.resolve(directory, process.env.NETLD_OUTPUT_FILE || `inventory.${format}`),
    format,
    pageSize,
    scheme: process.env.NETLD_SEARCH_SCHEME?.trim() || "ipAddress",
    query: process.env.NETLD_SEARCH_QUERY || "",
  };
}

class NetLDClient {
  constructor(baseUrl, apiKey, timeout = 30000) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.timeout = timeout;
    this.cookies = "";
  }

  headers(extra = {}) {
    return { Authorization: `Bearer ${this.apiKey}`, ...(this.cookies ? { Cookie: this.cookies } : {}), ...extra };
  }

  async request(options = {}) {
    let response;
    try {
      response = await fetch(`${this.baseUrl}/rest`, {
        redirect: "manual", signal: AbortSignal.timeout(this.timeout), ...options,
        headers: this.headers(options.headers),
      });
    } catch {
      throw new ExampleError(`Could not reach ${this.baseUrl}.`);
    }
    if (response.status >= 300 && response.status < 400) {
      throw new ExampleError(`Request redirected to ${response.headers.get("location") || ""}.`);
    }
    if (!response.ok) throw new ExampleError(`Request failed with HTTP ${response.status}.`);
    const setCookies = response.headers.getSetCookie?.() || splitSetCookie(response.headers.get("set-cookie") || "");
    if (setCookies.length) this.cookies = setCookies.map((item) => item.split(";", 1)[0]).join("; ");
    return response;
  }

  async login() {
    await this.request();
  }

  async call(method, parameters = {}) {
    const response = await this.request({
      method: "POST", headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ jsonrpc: "2.0", method, params: parameters, id: randomUUID() }),
    });
    const data = await response.json();
    if (data.error) throw new ExampleError(`${method} failed: ${JSON.stringify(data.error)}`);
    return data.result;
  }

  async searchInventory(networks, scheme, query, offset, pageSize) {
    const result = await this.call("Inventory.search", {
      network: networks, scheme, query: query.endsWith("\n") ? query : `${query}\n`,
      pageData: { offset, pageSize }, sortColumn: "ipAddress", descending: false,
    });
    if (!result || typeof result !== "object") throw new ExampleError("Inventory.search returned no page data.");
    return result;
  }
}

function splitSetCookie(header) {
  return header ? header.split(/,(?=\s*[^;=]+=[^;]+)/) : [];
}

export async function exportInventory(client, options) {
  const { networks, scheme, query, pageSize, outputFile, format = "csv" } = options;
  if (!["csv", "json"].includes(format)) throw new ExampleError("Output format must be either csv or json.");
  await mkdir(path.dirname(outputFile), { recursive: true });
  const temporaryFile = path.join(path.dirname(outputFile), `.${path.basename(outputFile)}.${randomUUID()}.tmp`);
  let handle;
  let count = 0;
  try {
    handle = await open(temporaryFile, "wx");
    if (format === "csv") await handle.write(csvRow(CSV_FIELDS));
    else await handle.write("[\n");
    let firstJsonRecord = true;
    let offset = 0;
    let total;
    while (true) {
      const page = await client.searchInventory(networks, scheme, query, offset, pageSize);
      const devices = page.devices || [];
      if (!Array.isArray(devices)) throw new ExampleError("Inventory.search returned an invalid devices collection.");
      for (const device of devices) {
        if (format === "csv") {
          await handle.write(csvRow(CSV_FIELDS.map((field) => device[field])));
        } else {
          const record = Object.fromEntries(CSV_FIELDS.map((field) => [field, device[field] ?? null]));
          if (!firstJsonRecord) await handle.write(",\n");
          const rendered = JSON.stringify(record, null, 2).split("\n").map((line) => `  ${line}`).join("\n");
          await handle.write(rendered);
          firstJsonRecord = false;
        }
        count += 1;
      }
      const returnedPageSize = Number(page.pageSize || pageSize);
      if (!Number.isInteger(returnedPageSize) || returnedPageSize <= 0) {
        throw new ExampleError("Inventory.search returned an invalid page size.");
      }
      if (total == null && page.total != null) total = Number(page.total);
      if (total != null && offset + devices.length >= total) break;
      if (total == null && devices.length < returnedPageSize) break;
      if (!devices.length) throw new ExampleError("Inventory.search returned an empty page before the reported total.");
      offset += returnedPageSize;
    }
    if (format === "json") await handle.write("\n]\n");
    await handle.close();
    handle = undefined;
    await rename(temporaryFile, outputFile);
    return count;
  } catch (error) {
    await handle?.close().catch(() => {});
    await rm(temporaryFile, { force: true }).catch(() => {});
    throw error;
  }
}

export async function main() {
  const directory = path.dirname(fileURLToPath(import.meta.url));
  const envOffset = process.argv.indexOf("--env");
  const formatOffset = process.argv.indexOf("--format");
  if (envOffset >= 0 && !process.argv[envOffset + 1]) throw new ExampleError("--env requires a file path.");
  if (formatOffset >= 0 && !process.argv[formatOffset + 1]) throw new ExampleError("--format requires csv or json.");
  const envFile = envOffset >= 0 ? path.resolve(process.argv[envOffset + 1]) : path.join(directory, ".env");
  const config = await loadConfig(directory, envFile, formatOffset >= 0 ? process.argv[formatOffset + 1] : undefined);
  const client = new NetLDClient(config.baseUrl, config.apiKey);
  await client.login();
  const count = await exportInventory(client, config);
  console.log(`Wrote ${count} devices to ${config.outputFile}`);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch((error) => {
    console.error(`Error: ${error.message}`);
    process.exitCode = 1;
  });
}
